haskell-lsp-types/src/Language/Haskell/LSP/Types/Uri.hs

```diff
@@ -113,3 +142,48 @@ platformAdjustToUriPath systemOS srcPath
     unescaped c
       | systemOS == windowsOS = isUnreserved c || c `elem` [':', '\\', '/']
       | otherwise = isUnreserved c || c == '/'
+
+-- | Newtype wrapper around FilePath that always has normalized slashes.
+-- The NormalizedUri and hash of the FilePath are cached to avoided
+-- repeated normalisation when we need to compute them (which is a lot).
+--
+-- This is one of the most performance critical parts of ghcide, do not
+-- modify it without profiling.
+data NormalizedFilePath = NormalizedFilePath NormalizedUri !FilePath
+    deriving (Generic, Eq, Ord)
+
+instance NFData NormalizedFilePath
+
+instance Binary NormalizedFilePath where
+  put (NormalizedFilePath _ fp) = put fp
+  get = do
+    v <- Data.Binary.get :: Get FilePath
+    return (toNormalizedFilePath v)
+
+instance Show NormalizedFilePath where
+  show (NormalizedFilePath _ fp) = "NormalizedFilePath " ++ show fp
+
+instance Hashable NormalizedFilePath where
+  hash (NormalizedFilePath uri _) = hash uri
+
+instance IsString NormalizedFilePath where
+    fromString = toNormalizedFilePath
+
+toNormalizedFilePath :: FilePath -> NormalizedFilePath
+toNormalizedFilePath fp = NormalizedFilePath nuri nfp
+  where nfp | fp == "" = "" 
+            -- ghcide want to keep empty paths instead of normalising them to "."
```

Comment:
@cocreature what do you think about remove this corner case? i would prefer to keep it in ghcide instead the lib